users/newgrp.rs

```diff
@@ -0,0 +1,764 @@
+    let loginname = get_current_login().unwrap_or("???".to_string());
+
+    // Get the current tty device name
+    let tty = get_current_tty().unwrap_or("???".to_string());
+
+    // Log the message
+    eprintln!(
+        "user '{}' (login '{}' on {}) switched to group with id '{}'",
+        name, loginname, tty, gid
+    );
+}
+
+/// Retrieves the current login name.
+fn get_current_login() -> Option<String> {
+    unsafe {
+        let login_ptr = libc::getlogin();
+        if !login_ptr.is_null() {
+            CStr::from_ptr(login_ptr)
+                .to_str()
+                .ok()
+                .map(|s| s.to_string())
+        } else {
+            None
+        }
+    }
+}
+
+/// Retrieves the current tty name.
+fn get_current_tty() -> Option<String> {
+    unsafe {
+        let tty_ptr = libc::ttyname(0);
+        if !tty_ptr.is_null() {
+            CStr::from_ptr(tty_ptr).to_str().ok().map(|s| {
+                if s.starts_with("/dev/") {
+                    s[5..].to_string()
+                } else {
+                    s.to_string()
+                }
+            })
+        } else {
+            None
+        }
+    }
+}
+
+/// Checks permissions for accessing a specified group based on the provided user credentials.
+///
+/// This function determines whether the user associated with the given password structure
+/// has permission to access the specified group. If a password is required, it prompts the
+/// user for the password and verifies it against the group's password.
+///
+/// # Arguments
+///
+/// * `group` - A reference to the `Group` struct representing the group whose permissions are being checked.
+/// * `password` - A `passwd` struct containing the user's password information.
+///
+/// # Returns
+///
+/// Returns `Ok(group.gid)` if the user has the necessary permissions to access the group. If a password is required
+/// and does not match, it returns an `Err` with `io::ErrorKind::PermissionDenied`.
+
+fn check_perms(group: &Group, password: passwd) -> Result<u32, io::Error> {
+    let pw_name = unsafe {
+        CStr::from_ptr(password.pw_name)
+            .to_string_lossy()
+            .into_owned()
+    };
+
+    // Determine if a password is needed based on group membership and GID
+    let mut need_password =
+        group.gid != password.pw_gid && group.members.iter().all(|member| member != &pw_name);
+
+    // Convert C-style strings (char pointers) to Rust &CStr and check for empty passwords
+    unsafe {
+        let user_password = CStr::from_ptr(password.pw_passwd).to_bytes();
+
+        if user_password.is_empty() && !group.passwd.is_empty() {
+            need_password = true;
+        }
+    }
+
+    // Check for permissions if necessary
+    unsafe {
+        if getuid() != 0 && need_password {
+            let password_input = read_password().unwrap_or_default();
+            let hashed_input = pw_encrypt(&password_input, Some(&group.passwd)).unwrap();
```

Comment:
Only extracts from /etc/group file. Which in some cases will return `x` if the actual password is in the `/etc/gshadow` file.